src/session/session_api.c

```diff
@@ -1423,6 +1435,37 @@ __open_session(WT_CONNECTION_IMPL *conn,
 	session_ret->id = i;
 	session_ret->iface = stds;
 	session_ret->iface.connection = &conn->iface;
+	/*
+	 * Disable some methods if this is a read-only connection.  Group
+	 * them by call signature.
+	 */
+	if (F_ISSET(conn, WT_CONN_READONLY)) {
+		wt_session = &session_ret->iface;
+		wt_session->checkpoint =
+		    (int (*)(WT_SESSION *, const char *))__wt_session_notsup;
+		wt_session->compact =
+		    (int (*)(WT_SESSION *, const char *, const char *))__wt_session_notsup;
```

Comment:
> as long as the compiler pops based on the caller's arguments, which I think it has to do

That's exactly my concern. Stackoverflow seems to answer my question:
http://stackoverflow.com/questions/559581/casting-a-function-pointer-to-another-type

tl;dr The behavior is undefined in the C standard, but likely to work most places. I don't feel strongly about it, it's definitely cleaner to have less duplication of the not supported functions.